cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/SourceControlManager.java

```diff
@@ -0,0 +1,244 @@
+   * @throws AuthenticationException when correct authentication credentials are not available.
+   * @throws IOException             when file or network I/O fails.
+   */
+  public void initialise() throws GitAPIException, AuthenticationException, IOException {
+    if (git != null) {
+      return;
+    }
+    // Clean up the directory if it already exists.
+    if (localRepoDirectory.exists() && !deleteDirectory(localRepoDirectory)) {
+      throw new IOException("Failed to clean up local repository directory " + context.getRepoBasePath());
+    }
+    String branch = "refs/heads/" + context.getRepositoryConfig().getDefaultBranch();
+    CloneCommand command = Git.cloneRepository()
+      .setURI(context.getRepositoryConfig().getDefaultBranch())
+      .setDirectory(localRepoDirectory)
+      .setBranchesToClone(Collections.singleton((branch)))
+      .setBranch(branch);
+
+    command = getAuthenticatedCommandWithTimeout(command);
+    command.call();
+  }
+
+  /**
+   * Returns an {@link AuthStrategy} for the given Git repository provider and auth type.
+   *
+   * @return an instance of {@link  AuthStrategy}.
+   * @throws AuthStrategyNotFoundException when the corresponding {@link AuthStrategy} is not found.
+   */
+  private AuthStrategy getAuthStrategy() throws AuthStrategyNotFoundException {
+    RepositoryConfig config = this.context.getRepositoryConfig();
+    if (config.getProvider() == Provider.GITHUB) {
+      if (config.getAuth().getType() == AuthType.PAT) {
+        return new GitPATAuthStrategy(this.context.getStore());
+      }
+    }
+    throw new AuthStrategyNotFoundException(String.format("No strategy found for provider %s and type %s.",
+                                                          config.getProvider(),
+                                                          config.getAuth().getType()));
+  }
+
+  /**
+   * Validated the configuration provided to this manager.
+   *
+   * @return a {@link List} of {@link  Exception}s occurred during validation.
+   */
+  public List<Exception> validateConfig() {
+    ArrayList<Exception> validationErrors = new ArrayList<>();
+    // Validate existence of secret in Secure Store.
+    RepositoryConfig config = context.getRepositoryConfig();
+    LOG.info("Checking existence of secret in secure store: " + config.getAuth().getTokenName());
+    try {
+      context.getStore().get(context.getNamespaceId().getNamespace(), config.getAuth().getTokenName()).get();
+    } catch (Exception e) {
+      validationErrors.add(new Exception("Failed to get password/token from secure store", e));
+      return validationErrors;
+    }
+
+    // Try fetching heads in the remote repository.
+    FileRepository localRepo = null;
+    if (validationRepoDirectory.exists() && !deleteDirectory(validationRepoDirectory)) {
+      validationErrors.add(new Exception(String.format("Failed to clean up local directory '%s' for temp repository",
```

Comment:
As a followup we should think about differentiating FS/Store failures and actual config issues in UI. Otherwise it can confuse the users.

Reply:
The FS exception is now thrown as an IO exception while the actual validation errors are returned as a RepositoryValidationFailure object now.